pilot/pkg/networking/core/v1alpha3/cluster_test.go

```diff
@@ -896,14 +876,28 @@ func TestConditionallyConvertToIstioMtls(t *testing.T) {
 				SubjectAltNames:   []string{"spiffee://foo/serviceaccount/1"},
 				Sni:               "foo.com",
 			},
+			autoDetected,
+		},
+		{
+			"Do not auto fill nil settings for external",
```

Comment:
Why are we using MeshExternal field for this? I thought the logic for auto mtls was if it has the label its mtls enabled, if not then it isn't. Where does internal/external have to do with this? It can be internal without mtls or external with mtls.

See https://github.com/istio/api/pull/1104, currently there are no usages of MESH_EXTERNAL

Reply:
+1 I don't know the full history of the MESH_EXTERNAL, but Iast time I checked with @louiscryan vaguely remembered the reason to introduce this field no longer exist (something related to mTLS before authn policy exists? so we use it to not turn on mTLS for external service before, @diemtvu  do you know?). We don't have any code base to reference this enum actually.

We should do a doc update follow up and deprecate it from our API.

Reply:
So mesh external was created precisely for this purpose :) [in the days before where we had a single global flag to enable/disable mTLS]. If I declare a google.com service entry and add the security.istio.io/mtlsReady: true, will we initiate a mTLS connection ? We shouldn't. Most users wouldn't either, but it would be helpful to explicitly forbid that or ignore mesh external so that people dont get caught with these issues. [This is just like how we do sni matching for ports of type https - usability improvements].

We also filter the set of services exposed in the buildOutboundSniDnatClusters to just the internal ones [not mesh external ones]. So there are two solid use cases of internal vs external.

Reply:
> If I declare a google.com service entry and add the security.istio.io/mtlsReady: true, will we initiate a mTLS connection ? We shouldn't.

Why not? If the user tells me to send mtls traffic I don't see why we shouldn't respect that. Its not like they need to explicitly add `security.istio.io/mtlsReady: false` to get things working, right? The default is already false, so they have to explicitly declare they are mtls ready.

Making use of MESH_EXTERNAL is roughly the same as if they field never existed and was added today, given we will hide it if not for this PR. I think if we this PR introduced a whole new field to specify location just for this we would not think it was worth it

Reply:
the field exists for two purposes:

one for the multi-cluster use case
> We also filter the set of services exposed in the buildOutboundSniDnatClusters to just the internal ones [not mesh external ones]. So there are two solid use cases of internal vs external.

and second for this auto mTLS thing.. adding a label that magically turns on an istio internal feature like mTLS when talking to external services seems like a bad idea IMO. If the user really wanted to do mTLs, she could declare that google.com as an internal service right?

Reply:
@rshriram I don't see where we are using it in buildOutboundSniDnatClusters, are you sure its in use still? Last I looked it was not except in some /debug endpoint.

Reply:
https://github.com/istio/istio/pull/17621/files#diff-f507fb55b7edf4a03d3c7e9aebf64a1eR306

Reply:
I dont think we have ever advertised that it is not necessary. The autoMTLS is new feature - i.e. nothing to break as we require endpoints to have the labels for it to be active. The sni-dnat could never work with external service. So, I dont think it would break anything.

Reply:
I think the point is yes, relying on mesh external won't break anything, but if we don't rely on it auto MTLS should also be working, you configure transport_socket_match with a EDS generated by ServiceEntry of "bing.com" would also work, fallback to the plaintext socket config.

We have't advertised that's not necessary but the fact our code base does not have anywhere to read from the field makes it not necessary per se. It's better to reduce user manual config if we can avoid.

Reply:
I missed previous discussion, I think we don't really need another INTERNAL/EXTERNAL one field to cover misconfiguration of another (adding label of google.com to be mtls ready)

Anyway, we can revisit this as a follow-up...